### DISCUSSION TITLE
Issue with AsyncLocalStorage in Next.js 14: context is not being shared between middleware, handlers and SSR pages

### DISCUSSION THREAD

**Discussion by joaquinbentancor** [vercel/next.js#67305] [Help]
### Summary

I am experiencing issues with sharing the AsyncLocalStorage context between Middleware, API Handlers, and SSR Pages in a Next.js 14 project using AsyncLocalStorage. Even though the context is correctly set in the middleware, it does not propagate to API handlers or server-rendered pages. Is this a expected behaviour? I want to achieve a custom distributed tracing solution that will be defined at middleware level and should be accesible in api handlers and ssr pages.
This works fine with a custom server with Express, but I would like to use the built-in server.

Here is a simplified example that reproduces the problem:

```
// utils/asyncLocalStorage.ts
import { AsyncLocalStorage } from 'async_hooks';

const asyncLocalStorage = new AsyncLocalStorage();
export default asyncLocalStorage;
```

```
// middleware.ts
import { NextRequest, NextResponse } from 'next/server';
import asyncLocalStorage from './utils/asyncLocalStorage';

export function middleware(req: NextRequest) {
  return asyncLocalStorage.run({}, async () => {
    const store = asyncLocalStorage.getStore();
    if (store) {
      store.key1 = test;
    }
    return NextResponse.next();
  });
}
```

```
// pages/api/hello.ts
import type { NextApiRequest, NextApiResponse } from 'next';
import asyncLocalStorage from '../../utils/asyncLocalStorage';

export default function handler(req: NextApiRequest, res: NextApiResponse) {
  const store = asyncLocalStorage.getStore();
  console.log(store); //Here the store is undefined
  res.status(200).send();
}
```

**Comment by MrDelik** [comment]
unfortunately, i have the same problem.
Did you manage to find a solution ?

**Comment by cklat** [comment]
Having the same issue and wondering about a solution particular to this problem or a general solution in integrating a proper logging mechanism with context propagation for a NextJs App.

My files look the following and I'm having trouble accessing the store server side when a transaction is triggered through the client by calling a route in the ApiRouter.

So the following setup for me as follows:

```
// ./utils/contextLogger.ts

import { AsyncLocalStorage } from 'async_hooks';
import { logger } from '@/next-logger.config';

interface LoggerContext {
  trace: string;
}

const asyncLocalStorage = new AsyncLocalStorage<LoggerContext>();

const getLoggerWithContext = () => {
  const context = asyncLocalStorage.getStore();

  if (context) {
    return logger({
      trace: context.trace,
    });
  }

  return logger;
};

const runWithAsyncContext = (fn: () => Promise<any>, context: LoggerContext) => {
  return asyncLocalStorage.run(context, fn);
};

export { getLoggerWithContext, runWithAsyncContext };
```

Example usage in Api Route:

```
// ./api/route.ts

import { NextResponse, NextRequest } from 'next/server';
import { getLoggerWithContext, runWithAsyncContext } from '@/utils/contextLogger';
import { v4 as uuidv4 } from 'uuid';

export const GET = (req: NextRequest) => {
  const trace = uuidv4();

  return runWithAsyncContext(async () => {
    const logger = getLoggerWithContext();
    logger.info('Info log');

    try {
      const s3EntryPoint = process.env.S3_ENTRYPOINT;
      const response = NextResponse.json({ s3EntryPoint });
      response.headers.set('Cache-Control', 'no-store');
      if(s3EntryPoint) {
        logger.info(`Another Info Log`)
      } else {
        logger.warn('A warn log.')
      }
      
      return response;
    } catch (error) {
      logger.error({ error }, 'An error log');
      return NextResponse.json({ error: 'Internal Server Error' }, { status: 500 });
    }
  }, { trace });
};

export const dynamic = "force-dynamic";
```

Then Server Side like so which is part of the transaction triggered through the call of the route above:

```
// .src/lib/backend.ts

import { getLoggerWithContext } from '@/utils/contextLogger';

export const fetchSomething = async (input: string): Promise<Data[]> => {
  
   /....

  const logger = getLoggerWithContext();
  
  logger.info('Fetching data')
  
  return /...
}
```

Here the logger.info will not populate the trace field with the UUID set in the route. This is because `asyncLocalStorage ` is not the same instance and gets newly created.

Any idea how to solve this or any other suggestions for a lightweight solution?

**Reply by janpolacek** [reply]
Have you found any solution?

**Reply by tuywen** [reply]
yes, i do it !! thanks very much.

i add two file

src/instrumentation.node.ts
```nodejs
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { NodeSDK } from '@opentelemetry/sdk-node';
import { SimpleSpanProcessor } from '@opentelemetry/sdk-trace-node';

const sdk = new NodeSDK({
  spanProcessor: new SimpleSpanProcessor(new OTLPTraceExporter()),
});

sdk.start();
```
src/instrumentation.ts`
```export async function register() {
  if (process.env.NEXT_RUNTIME === 'nodejs') {
    await import('./instrumentation.node');
  }
}
```

in my winston logger
```nodejs
try {
      const span = trace.getActiveSpan();
      const spanContext = span?.spanContext();
      return {
        traceId: spanContext?.traceId,
        spanId: spanContext?.spanId,
        traceFlags: spanContext?.traceFlags,
      };
    } catch (error) {
      return {};
    }
```

traceId works!

**Reply by tuywen** [reply]
@ivandamyanov  thanks very much

**Comment by talhayoun** [comment]
Any news regarding this?

**Comment by ivandamyanov** [comment]
Not sure if this is really the same but have you looked at server side instrumentation https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation and the new client side instrumentation: https://nextjs.org/docs/app/api-reference/file-conventions/instrumentation-client ?

**Comment by ffma-glen-sampson** [comment]
same problem

**Comment by sergey-hippo** [comment]
I'm using Next `15.5.4` with Pages router, with Next running inside custom Express server like this:
```ts
// server/index.ts
const app = next();
const handle = app.getRequestHandler();

....

app.prepare().then(() => {
  const server = express();
  ....
  
  server.get('*', async (req, res) => {
    await handle(req, res);
  });
}
```
```ts
// utils/asyncLocalStorage.ts
import { AsyncLocalStorage } from 'async_hooks';

const asyncLocalStorage = new AsyncLocalStorage();
export default asyncLocalStorage;
```

I tried 2 ways to make `AsyncLocalStorage`:
1. Via middleware like in parent post.
2. Directly in the handler that delegates to Next both via `run` and `enterWith`
```ts
// server/index.ts

// variant with run
...
  server.get('*', async (req, res) => {
    await asyncLocalStorage.run({}, async () => {
       await handle(req, res);
    }
  });
}

// variant with enterWith
...
  server.get('*', async (req, res) => {
     asyncLocalStorage.enterWith({});
     await handle(req, res);
  });
```

In all of those cases I could see the local storage in other middleware, but couldn't see it in `getServerSideProps` or anything called from it.

```ts
// pages/my.tsx
...

export async function getServerSideProps() {
  const store = asyncLocalStorage.getStore();
  console.log('in getServerSideProps', { store }); // store is always undefined here
  
  ...

}

```

**Comment by sergey-hippo** [comment]
Related: https://github.com/vercel/next.js/discussions/76529

**Comment by sl33kr** [comment]
Based on this change in NextJS 16, it appears this wouldn't be an intended use of what is now `proxy.js`
https://nextjs.org/docs/app/api-reference/file-conventions/proxy

>Proxy is meant to be invoked separately of your render code and in optimized cases deployed to your CDN for fast redirect/rewrite handling, you should not attempt relying on shared modules or globals.****
